Subject: [PATCH] feature: flash_erase command, ported for busybox 1.36.1

Add the flash_erase command (START_OFFSET + BLOCK_COUNT arguments,
0 count = whole device), replacing flash_eraseall whose functionality
is covered by "flash_erase /dev/mtdX 0 0" (Curt Brune, ONIE).

Rebased onto the 1.36.1 flash_eraseall.c (maintained upstream);
registration via //config://applet://kbuild: comments.
diff --git a/miscutils/flash_eraseall.c b/miscutils/flash_erase.c
similarity index 68%
rename from miscutils/flash_eraseall.c
rename to miscutils/flash_erase.c
index c76d9a6..ecb6946 100644
--- a/miscutils/flash_eraseall.c
+++ b/miscutils/flash_erase.c
@@ -1,6 +1,6 @@
 /* vi: set sw=4 ts=4: */
 /*
- * eraseall.c -- erase the whole of a MTD device
+ * flash_erase.c -- erase blocks of a MTD device
  *
  * Ported to busybox from mtd-utils.
  *
@@ -9,25 +9,31 @@
  * Renamed to flash_eraseall.c
  *
  * Licensed under GPLv2 or later, see file LICENSE in this source tree.
+ *
+ * 4/2013 -- Curt Brune <curt@cumulusnetworks.com>
+ *   Renamed to flash_erase.c
+ *   Added flash_erase command.
+ *   Removed flash_eraseall command.
  */
-//config:config FLASH_ERASEALL
-//config:	bool "flash_eraseall (5.9 kb)"
+//config:config FLASH_ERASE
+//config:	bool "flash_erase"
 //config:	default n  # doesn't build on Ubuntu 8.04
 //config:	help
-//config:	The flash_eraseall binary from mtd-utils as of git head c4c6a59eb.
-//config:	This utility is used to erase the whole MTD device.
+//config:	The flash_erase binary based on mtd-utils as of git head
+//config:	c4c6a59eb. This utility is used to erase specific erase
+//config:	blocks within a MTD device.
 
-//applet:IF_FLASH_ERASEALL(APPLET(flash_eraseall, BB_DIR_USR_SBIN, BB_SUID_DROP))
+//applet:IF_FLASH_ERASE(APPLET(flash_erase, BB_DIR_USR_SBIN, BB_SUID_DROP))
 /* not NOEXEC: if flash operation stalls, use less memory in "hung" process */
 
-//kbuild:lib-$(CONFIG_FLASH_ERASEALL) += flash_eraseall.o
+//kbuild:lib-$(CONFIG_FLASH_ERASE) += flash_erase.o
 
-//usage:#define flash_eraseall_trivial_usage
-//usage:       "[-jNq] MTD_DEVICE"
-//usage:#define flash_eraseall_full_usage "\n\n"
-//usage:       "Erase an MTD device\n"
+//usage:#define flash_erase_trivial_usage
+//usage:       "[-jq] MTD_DEVICE START_OFFSET BLOCK_COUNT"
+//usage:#define flash_erase_full_usage "\n\n"
+//usage:       "Erase blocks of the specified MTD device.\n"
+//usage:       "Specify a count of 0 to erase to end of device.\n"
 //usage:     "\n	-j	Format the device for jffs2"
-//usage:     "\n	-N	Don't skip bad blocks"
 //usage:     "\n	-q	Don't display progress messages"
 
 #include "libbb.h"
@@ -62,30 +68,39 @@ but mtd/jffs2-user.h is gone now (at least 2.6.31.6 does not have it anymore)
 #define cpu_to_je16(v) ((jint16_t){(v)})
 #define cpu_to_je32(v) ((jint32_t){(v)})
 
-static void show_progress(mtd_info_t *meminfo, erase_info_t *erase)
+static void show_progress(unsigned erasesize, unsigned current, unsigned start, unsigned total)
 {
 	printf("\rErasing %u Kibyte @ %x - %2u%% complete.",
-		(unsigned)meminfo->erasesize / 1024,
-		erase->start,
-		(unsigned) ((unsigned long long) erase->start * 100 / meminfo->size)
+	       erasesize / 1024,
+	       current,
+	       (unsigned) (((unsigned long long)(current - start) * 100) / total)
 	);
 	fflush_all();
 }
 
-int flash_eraseall_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
-int flash_eraseall_main(int argc UNUSED_PARAM, char **argv)
+int flash_erase_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
+int flash_erase_main(int argc UNUSED_PARAM, char **argv)
 {
 	struct jffs2_unknown_node cleanmarker;
 	mtd_info_t meminfo;
 	int fd, clmpos, clmlen;
 	erase_info_t erase;
 	struct stat st;
-	unsigned int flags;
+	unsigned n_erase_bytes;
+	unsigned flags;
 	char *mtd_name;
+	unsigned start, n_blocks;
+
+	flags = getopt32(argv, "^" "jNq" "\0" "=3");
 
-	flags = getopt32(argv, "^" "jNq" "\0" "=1");
+	argv += optind;
+	// argv[0] - device
+	mtd_name = argv[0];
+	/* offset parameter */
+	start = xstrtoul(argv[1], 0);
+	/* n_blocks parameter */
+	n_blocks = xstrtoul(argv[2], 0);
 
-	mtd_name = argv[optind];
 	fd = xopen(mtd_name, O_RDWR);
 	fstat(fd, &st);
 	if (!S_ISCHR(st.st_mode))
@@ -96,6 +111,24 @@ int flash_eraseall_main(int argc UNUSED_PARAM, char **argv)
 	if (meminfo.type == MTD_NANDFLASH)
 		flags |= IS_NAND;
 
+	if (n_blocks == 0) {
+		/* erase entire device */
+		n_erase_bytes = meminfo.size;
+	} else {
+		n_erase_bytes = meminfo.erasesize * n_blocks;
+	}
+
+	/*
+	 * Sanitize inputs:
+	 *   - starting offset aligned to erasesize boundary
+	 *   - requested region is completely within device
+	 */
+	if (start % meminfo.erasesize)
+		bb_error_msg_and_die("%s: offset 0x%08x not aligned to erase size 0x%08x",
+				     mtd_name, start, meminfo.erasesize);
+	if ((start + n_erase_bytes) > meminfo.size)
+		bb_error_msg_and_die("%s: erase region is outside of device", mtd_name);
+
 	clmpos = 0;
 	clmlen = 8;
 	if (flags & OPTION_J) {
@@ -149,7 +182,7 @@ int flash_eraseall_main(int argc UNUSED_PARAM, char **argv)
 	/* Don't want to destroy progress indicator by bb_error_msg's */
 	applet_name = xasprintf("\n%s: %s", applet_name, mtd_name);
 
-	for (erase.start = 0; erase.start < meminfo.size;
+	for (erase.start = start; erase.start < (start + n_erase_bytes);
 	     erase.start += meminfo.erasesize) {
 		if (!(flags & OPTION_N)) {
 			int ret;
@@ -176,7 +209,7 @@ int flash_eraseall_main(int argc UNUSED_PARAM, char **argv)
 		}
 
 		if (!(flags & OPTION_Q))
-			show_progress(&meminfo, &erase);
+			show_progress(meminfo.erasesize, erase.start, start, n_erase_bytes);
 
 		xioctl(fd, MEMERASE, &erase);
 
@@ -208,7 +241,7 @@ int flash_eraseall_main(int argc UNUSED_PARAM, char **argv)
 			printf(" Cleanmarker written at %x.", erase.start);
 	}
 	if (!(flags & OPTION_Q)) {
-		show_progress(&meminfo, &erase);
+		show_progress(meminfo.erasesize, erase.start, start, n_erase_bytes);
 		bb_putchar('\n');
 	}
 
